feat(prototypes): mention scenario with dropdown

bmdalex opened the pull request:
## feat(prototypes): mention scenario with dropdown

### Description

This PR attempts to create a prototype to show users how they can mention people using `@` key in a rich text editor.

I am using a `contentEditable` div as rich text input and the `Dropdown` component to leverage the list of users and filtering through them. The rest of the functionality is in the prototype section 

### Screenshot
![screen recording 2019-02-14 at 16 51 35](https://user-images.githubusercontent.com/5442794/52798843-0421a280-3079-11e9-9aba-96f83624f778.gif)

### Issues:
- #900 and issue in [description](https://github.com/stardust-ui/react/pull/900#issue-253086393) - **blocker**
- styling regression when using the filtering capability caused by #679 

### Later edit:
all issues and blocking items for this PR are fixed

bmdalex committed:
proto(dropdown): @mention scenario

bmdalex committed:
reverted AsyncDropdownSearch changes

bmdalex commented on docs/src/prototypes/dropdowns/AsyncDropdownSearch.tsx:
```diff
@@ -34,25 +34,55 @@ const createEntry = (): Entry => ({
   }
 
-  fetchItems = () => {
+  private fetchItems = () => {
```
https://github.com/stardust-ui/react/pull/901#discussion_r256900108 from @layershifter :

_What is purpose of these changes? Do we need them?
We don't have an agreement on visibility modifiers so let's revert these changes._

bmdalex replied:
@layershifter 

> We don't have an agreement on visibility modifiers so let's revert these changes.

I don't get this part, it's pretty obvious to me that in Typescript class members that are not meant to be used outside the scope of the class are private.

> What is purpose of these changes? Do we need them?

I will revert the changes however because the prototype section does not have the JS version so you might argue the change does not bring any benefit here.

kuzhelov replied:
agree with both sides here. Actually, if we would be absolutely pedantic here and adhere to the flavour of OO paradigm introduced by TS, we indeed should use scope modifiers for components' methods - to showcase which ones meant to be accessible from outside as part of public API.

At the same time, the main objection around having this burden of scope modifiers was reasonable fact that this is pretty rare scenario (that, almost for sure would be invalid in Stardust's case) when we would need to access methods of instantiated React component from outside world. Given this reasonable fact, it was considered to be unnecessary to have this scope modifiers being provided to React components.

That being said, just for the sake of moving forward, lets stick to the thing we've agreed upon (i.e. not introduce scope modifiers for React components) - although, as @Bugaa92 has noticed, this is the prototype's code where this is reasonable approach. And, if there is still strong objection against that (or, for certain, if there is any use case that showcases importance of having these modifiers) - @Bugaa92, please, feel free to schedule some time to discuss that 👍

layershifter replied:
Personally for me: there is no big difference, but I _expect_ to see handlers and helpers method before `render()` methods as it's done in React docs: https://reactjs.org/docs/handling-events.html

bmdalex replied:
let's close for now and stick to what we have, tracking in #933

bmdalex commented on docs/src/prototypes/dropdowns/utils.ts:
```diff
@@ -0,0 +1,36 @@
+    elem.innerText = text
+  }
+  range.insertNode(elem)
+}
```
https://github.com/stardust-ui/react/pull/901#discussion_r256902962 from @layershifter 

Is there any third party dependency that does this work? That is fully tested and supported?
Why we can't use [`caret-pos`](https://www.npmjs.com/package/caret-pos)?

bmdalex replied:
@layershifter Just wanted to keep things minimal and avoid  introducing extra dependencies just for a prototype, clients using this code will end up needing it too.

bmdalex commented on docs/src/prototypes/dropdowns/inputWithDropdown.tsx:
```diff
@@ -0,0 +1,136 @@
+    insertNodeAtCursorPosition({ id: this.mountNodeId })
+
+    const node = this.getMountNode()
+    ReactDOM.render(
```
https://github.com/stardust-ui/react/pull/901#discussion_r256901115 from @layershifter 

Can you please why we can't use `createPortal()`/`PortalInner` component there? What is a reason for another React's root?

bmdalex replied:
no reason, 1st iteration, let me try that, thanks for the suggestion

bmdalex replied:
implemented using `createPortal`, works as expected, good call

bmdalex committed:
implemented the new Dropdown using ReactDOM.createPortal instead of ReactDOM.render

layershifter commented on docs/src/prototypes/dropdowns/dataMocks.ts:
```diff
@@ -0,0 +1,53 @@
+  content: string
+}
+
+export const atMentionItems: AtMention[] = [
```
```suggestion
export const atMentionItems: AtMentionItem[] = [
```

Type name is not represent the object, `AtMentionItem` should be more clear

bmdalex replied:
done, thanks

layershifter commented on docs/src/prototypes/dropdowns/dataMocks.ts:
```diff
@@ -0,0 +1,53 @@
+    header: 'Selina Kyle',
+    image: 'public/images/avatar/small/ade.jpg',
+    content: 'Graphic Designer',
+  },
```
Actually, we can use `faker` to create these data. Can we use it?

bmdalex replied:
done, thanks

layershifter commented on docs/src/prototypes/dropdowns/index.tsx:
```diff
@@ -0,0 +1,21 @@
+import * as React from 'react'
+import { PrototypeSection, ComponentPrototype } from '../Protoypes'
```
```suggestion
import { PrototypeSection, ComponentPrototype } from '../Prototypes'
```

Oops, a typo ⌨️

layershifter commented on docs/src/prototypes/dropdowns/inputWithDropdown.tsx:
```diff
@@ -0,0 +1,143 @@
+    padding: '5px',
+    minHeight: '100px',
+    outline: 0,
+  }
```
`editorStyle` can be a usual variable

bmdalex replied:
done, thanks

layershifter commented on docs/src/prototypes/dropdowns/inputWithDropdown.tsx:
```diff
@@ -0,0 +1,143 @@
+  }
+
+  private handleInputKeyDown = (e: React.KeyboardEvent) => {
+    const code = keyboardKey.getCode(e)
```
```suggestion
    const keyCode = keyboardKey.getCode(e)
```

bmdalex replied:
done, thanks

layershifter commented on docs/src/prototypes/dropdowns/inputWithDropdown.tsx:
```diff
@@ -0,0 +1,143 @@
+  private hideDropdownAndRestoreEditor = () => {
+    removeElement(this.getMountNode())
+
+    this.tryFocusEditor()
```
```ts
_.invoke(this.contendEditableRef.current, 'focus')
```

bmdalex replied:
done, thanks

layershifter commented on docs/src/prototypes/dropdowns/inputWithDropdown.tsx:
```diff
@@ -0,0 +1,143 @@
+        items={atMentionItems}
+        toggleIndicator={null}
+        searchInput={{
+          input: { autoFocus: true },
```
Can we avoid `setInputElementSize()` and just pass `size` as prop?

```ts
searchInput={{
  input: {
    autoFocus: true,
    size: dropdownOpen ? 0 : searchQuery.length.
  },
}}

bmdalex replied:
yup, with the portal approach it's possible, good call; when I was rendering another node it didn't work; also, size needs to be min 1 so I did `searchQuery.length + 1`

bmdalex committed:
- addressed PR comments;
- refactoring of using portal
- fix for all styling regressions

layershifter commented on docs/src/prototypes/Prototypes.tsx:
```diff
@@ -0,0 +1,29 @@
+  description?: string
+}
+
+export const PrototypeSection: React.SFC<ComponentPrototypeProps> = props => (
```
```suggestion
export const PrototypeSection: React.FC<ComponentPrototypeProps> = props => (
```

SFC was deprecated :)

bmdalex replied:
done, thanks

layershifter commented on docs/src/prototypes/Prototypes.tsx:
```diff
@@ -0,0 +1,29 @@
+  </Box>
+)
+
+export const ComponentPrototype: React.SFC<ComponentPrototypeProps> = props => (
```
```suggestion
export const ComponentPrototype: React.FC<ComponentPrototypeProps> = props => (
```

bmdalex replied:
done, thanks

layershifter commented on docs/src/prototypes/dropdowns/inputWithDropdown.tsx:
```diff
@@ -0,0 +1,109 @@
+  private readonly initialState: InputWithDropdownState = {
+    dropdownOpen: false,
+    searchQuery: '',
+  }
```
NIT: `initialState` can be extracted to a variable as it doesn't have any coupling with the class

bmdalex replied:
agreed on the fact there is no coupling, but the concept of `state` doesn't make sense outside of the component...

with that argument (`it doesn't have any coupling with the class`) most of `readonly` members in TS do not have any coupling with their class (are merely constants) but we add them as members because they make sense just in that class

layershifter commented on docs/src/prototypes/dropdowns/inputWithDropdown.tsx:
```diff
@@ -0,0 +1,109 @@
+          onKeyUp={this.handleEditorKeyUp}
+          style={editorStyle}
+        />
+        {this.renderDropdown()}
```
NIT: Do we really need `renderDropdown()`, it looks quite simple to be inlined

bmdalex replied:
with the latest refactoring..no, we don't need it, inlined it, thanks

layershifter commented on docs/src/prototypes/dropdowns/inputWithDropdown.tsx:
```diff
@@ -0,0 +1,109 @@
+
+  private hideDropdownAndRestoreEditor = (cb?: () => void) => {
+    this.setState(this.initialState, () => {
+      this.tryFocusEditor()
```
```suggestion
      _.invoke(this.contendEditableRef.current, 'focus')
```

NIT: `tryFocusEditor()`  used only once, may be we can remove it?

bmdalex replied:
I think this is a matter of personal preference; I prefer keeping it more verbose, it makes it obvious to the reader what the code does without to understand details.

layershifter commented on packages/react/src/components/Dropdown/DropdownSearchInput.tsx:
```diff
@@ -122,8 +124,8 @@ class DropdownSearchInput extends UIComponent<ReactProps<DropdownSearchInputProp
           onBlur: this.handleInputBlur,
           onKeyDown: this.handleInputKeyDown,
           ...accessibilityInputProps,
+          ...unhandledProps.input,
```
Looks weird, but I understand _why_ it's done.

bmdalex replied:
yea..no better idea at this point

kuzhelov replied:
share the same feeling as @layershifter expressed 😕

bmdalex committed:
another round of comments addressed

bmdalex committed:
renamed file because of case insensitive behavior on Mac

bmdalex committed:
- fixed bug with dropdown not being deleted from editor;
- fixed bug with text not being inserted when dropdown is closed;
- small refactoring of CustomPortal -> PortalAtCursorPosition

bmdalex committed:
improved documentation for itemToString prop

kuzhelov commented on docs/src/prototypes/dropdowns/AsyncDropdownSearch.tsx:
```diff
@@ -68,33 +68,26 @@ class AsyncDropdownSearch extends React.Component<{}, SearchPageState> {
-        </Segment>
-      </div>
+      <>
+        <Dropdown
```
for this experience might suggest to introduce
- comment to example's code
- maybe better additional `<p>` in between of `Dropdown` and `CodeSnippet`, or add text to `Divider`

that would explain that `CodeSnippet` part rendered is necessary to reflect state of `fetchedData` - otherwise it was a bit confusing to me an, frankly, originally I thought that this is just a docs bug :) In either case, it would be helpful to explicitly say to the client that this is aimed to represent state of _fetched_ data.

![image](https://user-images.githubusercontent.com/9024564/53235887-f2a54f80-36a3-11e9-8f98-497a58a1a4ee.png)

Another small suggestion might be about reducing amount of data entries fetched - I think that even 5 (with names that would suggest reasonably good distribution) should work fine :) Key benefit is that it won't be necessary for the client to make long scroll to get to the next example (if this data section would be collapsable, it would help as well).

kuzhelov replied:
also, when there is nothing found in the list, there is the following view of Dropdown produced:

![image](https://user-images.githubusercontent.com/9024564/53236133-9db60900-36a4-11e9-993e-3b93224f743f.png)

This representation (with narrow white ribbon below the search field) seems to be a bit confusing - I think that we could easily fix it with `noResultsMessage` prop:

![image](https://user-images.githubusercontent.com/9024564/53236285-f685a180-36a4-11e9-8c19-85a71565f4b7.png)

bmdalex replied:
@kuzhelov `AsyncDropdownSearch.tsx` is not part of this PR;
 this file was just moved so that all dropdown related prototypes are in one place

kuzhelov commented on docs/src/prototypes/dropdowns/MentionsWithDropdown.tsx:
```diff
@@ -0,0 +1,109 @@
+  outline: 0,
+}
+
+class MentionsWithDropdown extends React.Component<{}, MentionsWithDropdownState> {
```
this is just about visual representation - frankly, it was not absolutely evident to me that this grey box below is introduced as an _editable area_. 

![image](https://user-images.githubusercontent.com/9024564/53237213-3057a780-36a7-11e9-8f53-054bdf61405d.png)


I do understand that this was done to keep styles (and amount of code) minimal - and totally support that 💯 But maybe we could improve a bit in this aspect, by doing the following steps:

- change wording of description a bit to explicitly mention that area is editable
- use brighter styles that would differentiate area from regular rectangle

![image](https://user-images.githubusercontent.com/9024564/53237116-ef5f9300-36a6-11e9-8c6e-039bedd7ff03.png)


Please, feel free to decide on whether changes should be done, and which ones specifically. Let me just, for the sake of reference, share changes that I have tried on my side to achieve this appearance:

```js
const editorStyle: React.CSSProperties = {
  backgroundColor: '#eee',
  padding: '5px',
  minHeight: '100px',
  outline: 0,
  border: '1px dashed grey',
  borderRadius: '5px'
}
```

bmdalex replied:
I like it, applied the text change and style on content editable div, thanks!

kuzhelov commented on docs/src/routes.tsx:
```diff
@@ -62,9 +62,9 @@ const Router = () => (
-            key="/prototype-async-dropdown-search"
-            path="/prototype-async-dropdown-search"
-            component={require('./prototypes/AsyncDropdownSearch/index').default}
+            key="/prototype-dropdowns"
```
👍

kuzhelov commented on packages/react/src/components/Dropdown/Dropdown.tsx:
```diff
@@ -101,7 +101,10 @@ export interface DropdownProps extends UIComponentProps<DropdownProps, DropdownS
 
   /**
-   * Function to be passed to create string from selected item, if it's a shorthand object. Used when dropdown also has a search function.
+   * Function that converts an item to string. Used when dropdown has the search boolean prop set to true.
```
👍

layershifter commented on docs/src/prototypes/Prototypes.tsx:
```diff
@@ -0,0 +1,29 @@
+}
+
+interface ComponentPrototypeProps extends PrototypeSectionProps {
+  description?: string
```
```suggestion
  description?: React.ReactNode
```

bmdalex replied:
done, thx

layershifter commented on docs/src/prototypes/Prototypes.tsx:
```diff
@@ -0,0 +1,29 @@
+import { Box, Header, Segment } from '@stardust-ui/react'
+
+interface PrototypeSectionProps {
+  title?: string
```
```suggestion
  title?: React.ReactNode
```

Please allow to pass elements.

bmdalex replied:
done, thx

kuzhelov commented on docs/src/prototypes/dropdowns/MentionsWithDropdown.tsx:
```diff
@@ -0,0 +1,109 @@
+            items={atMentionItems}
+            toggleIndicator={null}
+            searchInput={{
+              input: { autoFocus: true, size: searchQuery.length + 1 },
```
seems that `size` prop won't work, as there is no `size` prop declared for `DropdownSearchInput` (or for the `Input`). Am I missing something? I think it would be safe for us just to omit this prop

bmdalex replied:
`size` is an HTML attribute:

_unsigned long: Returns / Sets the element's size attribute, containing visual size of the control. This value is in pixels unless the value of type is text or password, in which case, it is an integer number of characters. Applies only when type is set to text, search, tel, url, email, or password; otherwise it is ignored.
(see: https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement)_

We need it to resize the input while typing; useful for mentioning people in the middle of text:
![screen recording 2019-02-22 at 14 48 06](https://user-images.githubusercontent.com/5442794/53246441-fc817f80-36b0-11e9-8eef-a519272b298d.gif)

kuzhelov replied:
got it, thanks for explanation - actually, the fact that we are overlapping with DOM property's name seems a bit disturbing to me. But this is, definitely, a discussion aside from this PR.

kuzhelov commented on docs/src/prototypes/dropdowns/MentionsWithDropdown.tsx:
```diff
@@ -0,0 +1,109 @@
+  }
+
+  private handleEditorKeyUp = (e: React.KeyboardEvent) => {
+    if (!this.state.dropdownOpen && keyboardKey.getCode(e) === keyboardKey.AtSign) {
```
this condition will open dropdown even if `2` is typed - I would rather suggest to introduce simplest possible logic of checking for `@` symbol explicitly

bmdalex replied:
done

kuzhelov reviewed (commented):
nice work 👍 just couple of things that I would like to be addressed before merge (apart from minor comments:
- sizing for `input` slot of dropdown input (could we remove this `size` altogether?): https://github.com/stardust-ui/react/pull/931/files#r259296667
- at-mentions dropdown is opened on `2` character typed: https://github.com/stardust-ui/react/pull/931/files#r259297611

bmdalex committed:
addressed comments and fixed issue with creating empty text node

kuzhelov reviewed (approved):
please, just introduce changelog entry before merging 👍

kuzhelov-ms committed:
improve visual appearance of async loading example

bmdalex committed:
changelog
